### PR TITLE
Embedding related fixes

### LINKED ISSUE #2979: No EmbeddingBag Functional
Hi,

It would be nice to have an EmbeddingBag functional implementation.

Thanks,
Shawn

### LINKED ISSUE #3104: sparse argument in Embedding and EmbeddingBag is not documented
The docstring for these two layers does not mention about the `sparse=True/False` argument they receive during initialization.

### PR DESCRIPTION
**Pull request by elanmart**
This PR aims to 
* fix https://github.com/pytorch/pytorch/issues/2979 (adding functional `EmbeddingBag`
* fix https://github.com/pytorch/pytorch/issues/3104 (missing docstring for keyword argument in Embedding)
* fix a bug in `Embedding` docstring (calling `variable[0].zero_()` results in `RuntimeError`)
* convert `nn.EmbeddingBag` to new style `Function`.

I've moved the core logic out of `EmbeddingBag` `Module` to the functional interface, since we would either have to dulpicate behaviour of `forward`. I hope this decision makes sense.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -705,6 +707,85 @@ def embedding(input, embedding_matrix,
...
+            indices = indices.view(-1)
+
+    elif indices.dim() != 1:
+        raise ValueError("input has to be 1D or 2D Tensor,"
+                         " but got Tensor of dimension {}".format(indices.dim()))
+
+    if offsets is None:
+        raise ValueError("offsets has to be a 1D Tensor but got None")
+
+    return torch.nn.backends.thnn.backend.EmbeddingBag.apply(
```
Comment (apaszke) [inline torch/nn/functional.py]: It's simpler to do `_functions.thnn.EmbeddingBag`

File: torch/nn/_functions/thnn/sparse.py [hunk 2]
```diff
@@ -120,35 +120,37 @@ def backward(ctx, grad_output):
...
-    def __init__(self, max_norm, norm_type, scale_grad_by_freq, mode):
-        super(EmbeddingBag, self).__init__()
-        self.max_norm = max_norm
-        self.norm_type = norm_type
-        self.scale_grad_by_freq = scale_grad_by_freq
-        self._indices = None
-        assert mode is not None
+    @staticmethod
+    def _renorm(ctx, indices, weight, max_norm, norm_type):
+        # clone indices since LookupTable_renorm modifies it in-place
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: so the old version didn't clone indices?

File: torch/nn/_functions/thnn/sparse.py [hunk 3]
```diff
@@ -164,57 +166,60 @@ def forward(self, weight, indices, offsets):
...
-        self.save_for_backward(indices)
+        ctx.save_for_backward(indices)
 
         indices = indices.contiguous().view(-1)
         output = weight.new()
-        if self.max_norm is not None:
-            self._renorm(indices, weight)
+
+        if ctx.max_norm is not None:
+            cls._renorm(ctx, indices, weight, max_norm=max_norm, norm_type=norm_type)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: Hmm it's kind of ugly to put `_renorm` in an autograd function, since it's not differentiated 😕 It's not a blocker for this PR in any way, I'll try to think about a nicer solution

**Review by apaszke — COMMENTED** [review]
Looks good! A few minor comments

File: torch/nn/_functions/thnn/sparse.py (hunk 2 above)
Comment (elanmart) [reply torch/nn/_functions/thnn/sparse.py]: I believe so:
https://github.com/pytorch/pytorch/blob/a0ac72e84e2f14c42d1bfdd695c411c0f101487c/torch/nn/_functions/thnn/sparse.py#L173-L176

File: torch/nn/functional.py (hunk 1 above)
Comment (elanmart) [reply torch/nn/functional.py]: sorry, I got confused since `Embedding` also calls its function via `torch.nn`.

File: torch/nn/functional.py [hunk 4]
```diff
@@ -655,6 +655,7 @@ def embedding(input, embedding_matrix,
         norm_type (float, optional): The p of the p-norm to compute for the max_norm option
         scale_grad_by_freq (boolean, optional): if given, this will scale gradients by the frequency of
                                                 the words in the mini-batch.
+        sparse (boolean, optional): if True, gradient w.r.t. weight matrix will be a sparse tensor.
```
[the remark this answers is not in the document]
Comment (martinraison) [reply torch/nn/functional.py]: I'm not aware of the latest developments but I think at least SGD/Adagrad should work (with the usual momentum/weight decay caveats).

[the remark this answers is not in the document]
Comment (apaszke) [reply torch/nn/functional.py]: We should at least say under what conditions can this be used (i.e. restrict to leaves and mention which optimizers can handle sparse updates)

File: torch/nn/functional.py (hunk 1 above)
Comment (apaszke) [reply torch/nn/functional.py]: No worries. I think this is valid too, it's just unnecessarily long

File: torch/nn/functional.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elanmart) [reply torch/nn/functional.py]: Should we specify which optimizers support sparse grad? Someone will have to remember to change it when new support is introduced.

BTW it seems only `sgd` supports `cuda` and `cpu`, and `adagrad` supports `cpu` only

File: torch/nn/functional.py [hunk 5]
```diff
@@ -655,12 +655,20 @@ def embedding(input, embedding_matrix,
...
+                                    more details regarding sparse gradients.
 
     Shape:
         - Input: LongTensor `(N, W)`, N = mini-batch, W = number of indices to extract per mini-batch
         - Embedding_matrix: FloatTensor `(V, embedding_dim)`, V = maximum index + 1, embedding_dim = embedding size
         - Output: `(N, W, embedding_dim)`
 
+    Notes:
+        It is advised to only use `sparse=True` if `embedding_matrix` is a leaf Variable,
+        since some autograd functions may not propagate sparse gradients correctly.
```
Comment (apaszke) [inline torch/nn/functional.py]: can you please rephrase this? It's not as bad as it sounds. You should never end up with incorrect gradients, but you might get errors in backward

File: torch/nn/modules/sparse.py [hunk 6]
```diff
@@ -28,6 +31,11 @@ class Embedding(Module):
         - Input: LongTensor `(N, W)`, N = mini-batch, W = number of indices to extract per mini-batch
         - Output: `(N, W, embedding_dim)`
 
+    Notes:
+        Keep in mind that only a limited number of optimizers support
```
Comment (apaszke) [inline torch/nn/modules/sparse.py]: Please mention leaf Variables

Comment (elanmart) [reply torch/nn/modules/sparse.py]: Sure, but should we mention it in the `Module` docs too? Since Embedding takes care of creating (leaf) weights Variable.

Comment (apaszke) [reply torch/nn/modules/sparse.py]: Ah sorry. No, it doesn't apply to modules :)

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks Marcin!
